--- a/CMakeListsGlobal.cmake
+++ b/CMakeListsGlobal.cmake
@@ -47,7 +47,7 @@ include (${CMAKE_CURRENT_LIST_DIR}/cmake
 #### get repository tag and version ####
 ########################################
 
-include (${CMAKE_CURRENT_LIST_DIR}/cmake/get_version.cmake)
-IF(EXISTS "${CMAKE_BINARY_DIR}/version.cmake")
-  include (${CMAKE_BINARY_DIR}/version.cmake)
-ENDIF()
+#include (${CMAKE_CURRENT_LIST_DIR}/cmake/get_version.cmake)
+#IF(EXISTS "${CMAKE_BINARY_DIR}/version.cmake")
+#  include (${CMAKE_BINARY_DIR}/version.cmake)
+#ENDIF()
