--- a/node_build/builder.js
+++ b/node_build/builder.js
@@ -203,8 +203,8 @@
     // # 74 "./wire/Message.h"
     js = js.replace(/\n#.*\n/g, '');
     var to = setTimeout(function () {
-        throw new Error("Inline JS did not return after 120 seconds [" + js + "]");
-    }, 120000);
+        throw new Error("Inline JS did not return after 5 minutes [" + js + "]");
+    }, 300000);

     var REQUIRE = function (str) {
         if (typeof(str) !== 'string') {
